nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,1115 @@
+           world coordinates.  Affine can also be None.  In this case,
+           ``obj.affine`` also returns None, and the affine as written to disk
+           will depend on the file format.
+        header : Cifti2Header object
+        nifti_header : None or mapping or nifti2 header instance, optional
+           metadata for this image format
+        '''
+        self._header = header or Cifti2Header()
+        self.data = data
+        self.extra = nifti_header
+
+    def get_data(self):
+        return self.data
+
+    @property
+    def shape(self):
+        return self.data.shape
+
+    @classmethod
+    def from_file_map(klass, file_map):
+        """ Load a Cifti2 image from a file_map
+
+        Parameters
+        file_map : file_map
+
+        Returns
+        -------
+        img : Cifti2Image
+            Returns a Cifti2Image
+         """
+        from .parse_cifti2_fast import _Cifti2AsNiftiImage, Cifti2Extension
+        nifti_img = _Cifti2AsNiftiImage.from_file_map(file_map)
+
+        # Get cifti2 header
+        for item in nifti_img.header.extensions:
+            if isinstance(item, Cifti2Extension):
+                cifti_header = item.get_content()
+                break
+        else:
+            raise ValueError('Nifti2 header does not contain a CIFTI2 '
+                             'extension')
+
+        # Construct cifti image
+        cifti_img = Cifti2Image(data=nifti_img.get_data()[0, 0, 0, 0],
+                                header=cifti_header,
+                                nifti_header=nifti_img.header)
+        cifti_img.file_map = nifti_img.file_map
+        return cifti_img
+
+    @classmethod
+    def from_image(klass, img):
+        ''' Class method to create new instance of own class from `img`
+
+        Parameters
+        ----------
+        img : ``spatialimage`` instance
+           In fact, an object with the API of ``FileBasedImage``.
+
+        Returns
+        -------
+        cimg : ``spatialimage`` instance
+           Image, of our own class
+        '''
+        if isinstance(img, klass):
+            return img
+        else:
+            raise NotImplementedError
+
+    def to_file_map(self, file_map=None):
+        """ Save the current image to the specified file_map
+
+        Parameters
+        ----------
+        file_map : string
+
+        Returns
+        -------
+        None
+        """
+        from .parse_cifti2_fast import Cifti2Extension
+        header = self.extra
+        extension = Cifti2Extension(content=self.header.to_xml())
+        header.extensions.append(extension)
+        data = np.reshape(self.data, (1, 1, 1, 1) + self.data.shape)
+        # If qform not set, reset pixdim values so Nifti2 does not complain
+        if header['qform_code'] == 0:
+            header['pixdim'][:4] = 1
+        img = Nifti2Image(data, None, header)
+        img.to_file_map(file_map or self.file_map)
+
+
+# class Cifti2DenseDataSeriesHeader(Cifti2Header):
```

Comment:
Why not break that off into a separate branch / PR as a reminder, and remove it here?